packages/next/client/head-manager.ts

```diff
@@ -58,6 +78,15 @@ function updateElements(
       if (title !== document.title) document.title = title
       return
     }
+    // handle react components as a function, functional component, or a class component
+    if (typeof tag.type === 'function') {
+      const newTags = reactComponentToDOM(tag)
+      newTags.forEach((nt: HTMLElement) => {
+        elements.add(nt)
+        headEl.appendChild(nt)
+      })
+      return
+    }
```

Comment:
so in my [first draft](https://github.com/vercel/next.js/pull/17770/commits/6601fa605408cc70ed3a1dc4a05a68f1b4f6b734#diff-e1a039509a8d92ebb1293f4f75a966cc981134c180b26f4422ed898862856b02R62), I just did an early return. That was a mistake as @timneutkens keenly pointed out cause it just throws away the rendering of the functional/class component on client-side navigation.

This is a hopeful fix because it now handles the function with or without children (array of elements).